### PR TITLE
fix(ecs): allow sub-org ecr repos

### PR DESCRIPTION
**Pull request by KyleJamesWalker**
The current regex match does not catch names like the following:
* `111111111111.dkr.ecr.us-east-1.amazonaws.com/sub-org/my-app-image:v0.0.0`

### CODE REVIEW

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcrImageProvider.java [hunk 1]
```diff
@@ -43,11 +43,11 @@
 @Component
 public class EcrImageProvider implements ImageRepositoryProvider {
   private static final Pattern ACCOUNT_ID_PATTERN = Pattern.compile("^([0-9]{12})");
-  private static final Pattern REPOSITORY_NAME_PATTERN = Pattern.compile("\\/([a-z0-9._-]+)");
+  private static final Pattern REPOSITORY_NAME_PATTERN = Pattern.compile("\\/([a-z0-9._\\/-]+)");
```
Comment (clareliguori) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcrImageProvider.java]: While you're at it, the official repo name pattern can be found here:
https://docs.aws.amazon.com/AmazonECR/latest/APIReference/API_CreateRepository.html#ECR-CreateRepository-request-repositoryName

It would be good to add negative unit tests, like repo name can't start with or end with a slash.

Comment (KyleJamesWalker) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcrImageProvider.java]: Thanks! I'll see if that documented regex works with the code as designed, and make some additional tests.

Comment (KyleJamesWalker) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcrImageProvider.java]: Increased tests to make sure the reference regex is working as expected, but haven't looked at adding some extra logic to cover the other possible edge cases you mentioned.

Comment (KyleJamesWalker) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcrImageProvider.java]: I was trying to do this fully within regex, but the max length check doesn't seem to be working, and I'm wondering if trying to add it is making the regex a little to difficult to quickly see what's going on. Here's my attempt to do the range check, if you have any ideas on why the max length isn't working now that you've had your coffee (I hope).

``` diff
diff --git a/clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcrImageProvider.java b/clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcrImageProvider.java
index d265ecfad..d6f5d0b6d 100644
--- a/clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcrImageProvider.java
+++ b/clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcrImageProvider.java
@@ -43,7 +43,7 @@ import java.util.stream.Collectors;
 @Component
 public class EcrImageProvider implements ImageRepositoryProvider {
   private static final Pattern ACCOUNT_ID_PATTERN = Pattern.compile("^([0-9]{12})");
-  private static final Pattern REPOSITORY_NAME_PATTERN = Pattern.compile("\\/((?:[a-z0-9]+(?:[._-][a-z0-9]+)*\\/)*[a-z0-9]+(?:[._-][a-z0-9]+)*)");
+  private static final Pattern REPOSITORY_NAME_PATTERN = Pattern.compile("\\/(((?:[a-z0-9]+(?:[._-][a-z0-9]+)*\\/)*[a-z0-9]+(?:[._-][a-z0-9]+)*){2,256})(?=[:@])");
   private static final String IDENTIFIER_PATTERN = "(:([a-z0-9._-]+)|@(sha256:[0-9a-f]{64}))";
   private static final Pattern REGION_PATTERN = Pattern.compile("(\\w+-\\w+-\\d+)");
   static final Pattern ECR_REPOSITORY_URI_PATTERN = Pattern.compile(ACCOUNT_ID_PATTERN.toString() + "\\.dkr\\.ecr\\." +
diff --git a/clouddriver-ecs/src/test/groovy/com/netflix/spinnaker/clouddriver/ecs/view/EcrImageProviderSpec.groovy b/clouddriver-ecs/src/test/groovy/com/netflix/spinnaker/clouddriver/ecs/view/EcrImageProviderSpec.groovy
index 878b55288..bd6d48afd 100644
--- a/clouddriver-ecs/src/test/groovy/com/netflix/spinnaker/clouddriver/ecs/view/EcrImageProviderSpec.groovy
+++ b/clouddriver-ecs/src/test/groovy/com/netflix/spinnaker/clouddriver/ecs/view/EcrImageProviderSpec.groovy
@@ -79,4 +79,44 @@ class EcrImageProviderSpec extends Specification {
     then:
     !canHandle
   }
+
+  @Test
+  void shouldNotHandleShortEcrRepositoryUrl() {
+    given:
+    EcrImageProvider ecrImageProvider = new EcrImageProvider(null, null)
+    String ecrRepositoryUrl = "123456789012.dkr.ecr.us-west-2.amazonaws.com/n:latest"
+
+    when:
+    boolean canHandle = ecrImageProvider.handles(ecrRepositoryUrl)
+
+    then:
+    !canHandle
+  }
+
+  @Test
+  void shouldHandleValidLengthRepositoryUrl() {
+    given:
+    EcrImageProvider ecrImageProvider = new EcrImageProvider(null, null)
+    // Good repo name lengths
+    String minRepo = "aa"
+    String maxRepo = "a".multiply(256)
+    // Bad repo name lengths
+    String shortRepo = "a"
+    String longRepo = "a".multiply(257)
+
+    String ecrRepositoryUrlMin = "123456789012.dkr.ecr.us-west-2.amazonaws.com/" + minRepo +":latest"
+    String ecrRepositoryUrlMax = "123456789012.dkr.ecr.us-west-2.amazonaws.com/" + maxRepo +":latest"
+    String ecrRepositoryUrlShort = "123456789012.dkr.ecr.us-west-2.amazonaws.com/" + shortRepo +":latest"
+    String ecrRepositoryUrlLong = "123456789012.dkr.ecr.us-west-2.amazonaws.com/" + longRepo +":latest"
+
+    when:
+    boolean canHandleMin = ecrImageProvider.handles(ecrRepositoryUrlMin)
+    boolean canHandleMax = ecrImageProvider.handles(ecrRepositoryUrlMax)
+    boolean canHandleShort = ecrImageProvider.handles(ecrRepositoryUrlShort)
+    boolean canHandleLong = ecrImageProvider.handles(ecrRepositoryUrlLong)
+
+    then:
+    canHandleMin && canHandleMax && !canHandleShort && !canHandleLong
+  }
+
 }```

Comment (clareliguori) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcrImageProvider.java]: I think it's fine to only do the non-empty check in regex and not do a max-length check.  What you have pushed works: this non-empty test passes against `Pattern.compile("\\/((?:[a-z0-9]+(?:[._-][a-z0-9]+)*\\/)*[a-z0-9]+(?:[._-][a-z0-9]+)*)");`, so this would be good to add to your changes

```
  @Test
  void shouldNotHandleMissingRepoName() {
    given:
    EcrImageProvider ecrImageProvider = new EcrImageProvider(null, null)
    String domainNameOnly = "123456789012.dkr.ecr.us-east-1.amazonaws.com"
    String noRepoName = "123456789012.dkr.ecr.us-east-1.amazonaws.com/"
    String subOrgNoRepoName = "123456789012.dkr.ecr.us-east-1.amazonaws.com/sub-org/"

    when:
    boolean canHandleDomainNameOnly = ecrImageProvider.handles(domainNameOnly)
    boolean canHandleNoRepoName = ecrImageProvider.handles(noRepoName)
    boolean canHandleSubOrgNoRepoName = ecrImageProvider.handles(subOrgNoRepoName)

    then:
    !canHandleDomainNameOnly && !canHandleNoRepoName && !canHandleSubOrgNoRepoName
  }
```

**Comment by KyleJamesWalker** [comment]
@spinnaker/reviewers
